sync/sync.go

```diff
@@ -0,0 +1,342 @@
+	LAYER_HASH p2p.MessageType = 2
+	LAYER_IDS  p2p.MessageType = 3
+
+	protocol = "/sync/fblock/1.0/"
+)
+
+func (s *Syncer) IsSynced() bool {
+	return s.layers.LocalLayerCount() == s.syncTo()
+}
+
+func (s *Syncer) Stop() {
+	s.exit <- true
+}
+
+func (s *Syncer) Start() {
+	if atomic.CompareAndSwapUint32(&s.startLock, 0, 1) {
+		go s.run()
+		return
+	}
+}
+
+func NewSync(peers Peers, layers mesh.Mesh, bv BlockValidator, conf Configuration) *Syncer {
+	s := Syncer{peers, layers, bv, conf, p2p.NewProtocol(peers, protocol, time.Second*5), 0, 0, make(chan bool), make(chan bool)}
+	s.p.RegisterMsgHandler(LAYER_HASH, s.LayerHashRequestHandler)
+	s.p.RegisterMsgHandler(BLOCK, s.BlockRequestHandler)
+	s.p.RegisterMsgHandler(LAYER_IDS, s.LayerIdsRequestHandler)
+	return &s
+}
+
+//fires a sync every sm.config.syncInterval or on force space from outside
+func (s *Syncer) run() {
+	syncTicker := time.NewTicker(s.config.syncInterval)
+	for {
+		doSync := false
+		select {
+		case <-syncTicker.C:
+			doSync = true
+		case doSync = <-s.forceSync:
+		case <-s.exit:
+			log.Debug("run stoped")
+			return
+		default:
+			doSync = false
+		}
+		if doSync {
+			go func() {
+				if atomic.CompareAndSwapUint32(&s.SyncLock, IDLE, RUNNING) {
+					s.Synchronise()
+					atomic.StoreUint32(&s.SyncLock, IDLE)
+				}
+			}()
+		}
+	}
+}
+
+func (s *Syncer) syncTo() uint32 {
+	if s.layers.LatestKnownLayer() < s.config.hdist {
+		return 0
+	}
+
+	return s.layers.LatestKnownLayer() - s.config.hdist
+}
+
+func (s *Syncer) Synchronise() {
+	for i := s.layers.LocalLayerCount(); i < s.syncTo(); {
+		i++
+		blockIds := s.GetLayerBlockIDs(i) //returns a set of all known blocks in the mesh
+		log.Debug("get layer block ids: ", blockIds)
+		ids := make(chan uint32, len(blockIds))
+		output := make(chan Block)
+
+		//bufferd chan for workers  to read from
+
+		for _, id := range blockIds {
+			ids <- id
+		}
+
+		close(ids)
+
+		// each worker goroutine tries to fetch a block iteratively from each peer
+		for i := 0; i < s.config.concurrency; i++ {
+			go func() {
+				for id := range ids {
+					for _, p := range s.peers.GetPeers() {
+						if bCh, err := s.SendBlockRequest(p, id); err == nil {
+							b := <-bCh
+							if s.sv.ValidateBlock(b) { //some validation testing
+								output <- b
+							}
+						}
+
+					}
+				}
+				close(output)
+			}()
+		}
+
+		blocks := make([]*mesh.Block, 0, len(blockIds))
+		for block := range output {
+			log.Debug("add block to layer", block)
+			blocks = append(blocks, mesh.NewExistingBlock(block.GetId(), block.GetLayer(), nil))
+		}
+		log.Debug("add layer ", i)
+		s.layers.AddLayer(mesh.NewExistingLayer(i, blocks))
+	}
+
+	log.Debug("synchronise done, local layer index is ", s.layers.LocalLayerCount())
+}
+
+func (s *Syncer) GetLayerBlockIDs(index uint32) []uint32 {
+
+	m := make(map[string]Peer, 20)
+	peers := s.peers.GetPeers()
+	// request hash from all
+
+	for _, p := range peers {
+		hash, err := s.SendLayerHashRequest(p, index)
```

Comment:
Why not make it concurrent?

Reply:
seemed like an overkill for this pr, ill open a task for it

Reply:
add a todo in code as well